src/engine.rs

```diff
@@ -397,6 +430,134 @@ impl Engine {
             .get_raw(account_id)
             .map(|result| Address(result.as_slice().try_into().unwrap()))
     }
+
+    pub fn register_token(&mut self, erc20_token: &[u8], nep141_token: &[u8]) {
+        // Check that this nep141 token was not registered before, they can only be registered once.
+        assert!(self.state.nep141_erc20.get_raw(nep141_token).is_none());
+
+        self.state
+            .erc20_nep141
+            .insert_raw(erc20_token, nep141_token);
+
+        self.state
+            .nep141_erc20
+            .insert_raw(nep141_token, erc20_token);
+    }
+
+    pub fn get_erc20_from_nep141(&self, nep141_token: &[u8]) -> Option<Vec<u8>> {
+        self.state.nep141_erc20.get_raw(nep141_token)
+    }
+
+    /// Transfers an amount from a given sender to a receiver, provided that
+    /// the have enough in their balance.
+    ///
+    /// If the sender can send, and the receiver can receive, then the transfer
+    /// will execute successfully.
+    pub fn transfer(
+        &mut self,
+        sender: Address,
+        receiver: Address,
+        value: U256,
+    ) -> EngineResult<SubmitResult> {
+        self.call(sender, receiver, value, Vec::new())
+    }
+
+    /// Mint tokens for recipient on a particular ERC20 token
+    /// This function should return the amount of tokens unused,
+    /// which will be always all (<amount>) if there is any problem
+    /// with the input, or 0 if tokens were minted successfully.
+    ///
+    /// The output will be serialized as a String
+    /// https://github.com/near/NEPs/discussions/146
+    ///
+    /// IMPORTANT: This function should not panic, otherwise it won't
+    /// be possible to return the tokens to the sender.
+    ///
+    /// TODO(#51): Carefully review that this function can't fail,
+    ///     and returns appropriate value
+    pub fn receive_erc20_tokens(&mut self) {
+        let input = sdk::read_input();
```

Comment:
I prefer the core logic being in `Engine`, and the IO happening in `lib.rs`. In this example

```rust
// engine.rs

pub fn receive_erc20_tokens(&mut self, sender_id: AccountId, amount: Balance, msg: String) -> Result<Balance, Error> { ... }
```

``` rust
// lib.rs

    /// Allow receiving NEP141 tokens to the EVM contract
    #[no_mangle]
    pub extern "C" fn ft_on_transfer() {
        #[allow(clippy::if_same_then_else)]
        if sdk::predecessor_account_id() == sdk::current_account_id() {
            // TODO(#59) ETH transfer
            todo!();
        } else {
            let mut engine = Engine::new(predecessor_address());
            let input = sdk::read_input();
            let result = match NEP141FtOnTransferArgs::try_from_slice(input.as_slice()) {
                Ok(args) => engine.receive_erc20_tokens(args.sender_id, args.amount, args.msg).ok_or(ags.amount),
                Err(_) => 0,
            }
            sdk::return_output(result.to_string.as_bytes());
        }
    }
```

The reason I think this is better is because 

1. It is more clear from the type signature what `receive_erc20_tokens` is trying to do
2. `receive_erc20_tokens` becomes easier to test (input is passed directly, reducing reliance on `sdk`)

Reply:
Please note that we are using the same approach for many functions (e.g. eth-connector related as well).
But I really like the suggestion. Especially the second item is really crucial.

Reply:
Yes, I know other eth-connector uses the style of reading input together with business logic. I think I mentioned this in a previous review of eth-connector code. I would like to see us move away from that pattern if possible though.

To be clear, I don't think this is a blocking concern (especially with timelines being what they are), but it would be nice to separate IO from logic where we can, and go back to fix all the places where we didn't in a future tech debt cycle.

Reply:
Totally agree with you!

Reply:
done!